draft-ietf-quic-recovery.md

```diff
@@ -645,14 +645,48 @@ packets might cause the sender's bytes in flight to exceed the congestion window
 until an acknowledgement is received that establishes loss or delivery of
 packets.
 
+## Persistent Congestion
+
 When an ACK frame is received that establishes loss of all in-flight packets
-sent prior to a threshold number of consecutive PTOs (pto_count is more than
-kPersistentCongestionThreshold, see {{cc-consts-of-interest}}), the network is
-considered to be experiencing persistent congestion, and the sender's congestion
-window MUST be reduced to the minimum congestion window (kMinimumWindow).  This
-response of collapsing the congestion window on persistent congestion is
-functionally similar to a sender's response on a Retransmission Timeout (RTO) in
-TCP {{RFC5681}}.
+sent over a long enough period of time, the network is considered to be
+experiencing persistent congestion.  Commonly, this can be established by
+consecutive PTOs, but since the PTO timer is reset when a new ack-eliciting
+packet is sent, an explicit duration must be used to account for those cases
+where PTOs do not occur or are substantially delayed.  This duration is the
+equivalent of kPersistentCongestionThreshold consecutive PTOS, and is computed
+as follows:
+~~~
+(smoothed_rtt + 4 * rttvar + max_ack_delay) *
+    ((2 ^ kPersistentCongestionThreshold) - 1)
+~~~
+
+For example, assume:
+
+  smoothed_rtt = 1
+  rttvar = 0
+  max_ack_delay = 0
+  kPersistentCongestionThreshold = 2
+
+If an eck-eliciting packet is sent at time = 0, the following scenario would
+illustrate persistent congestion:
+
+  t=0 | Send Pkt #1 (App Data)
+  t=1 | Send Pkt #2 (PTO 1)
+  t=3 | Send Pkt #3 (PTO 2)
+  t=7 | Send Pkt #4 (PTO 3)
+  t=8 | Recv ACK of Pkt #4
+
+The first three packets are determined to be lost when the ACK of packet 4 is
+received at t=8.  The congestion period is calculated as the time between the
+oldest and newest lost packets: (3 - 0) = 3.  The duration for persistent
```

Comment:
This doesn't match the description in the first paragraph, does it?
> When an ACK frame is received that establishes loss of all in-flight packets, sent over a long enough period of time, the network is considered to be experiencing persistent congestion.

With this algorithm, we would detect persistent congestion even if packet 2 and 4 were acked, right?

Reply:
The text says `all in-flight packets` meaning a contiguous block of sent packets. The example doesn't call out that check for a contiguous block explicitly, but it would have to be made. So, in that case if packets 2 and 4 were acknowledged you wouldn't have that contiguous block of time that meets the requirements.

Reply:
I think this example works, but the pseudocode below is not quite right for this reason.  The easiest way to fix it would be to assume no packet number gaps and check if the number of packets lost is equal to largest - smallest + 1.

Reply:
Do we need to really spell this out with a 100% complete pseudocode example? I can imagine a couple ways for implementations to achieve this and I don't think we need to call them all out.

Reply:
The text is normative and the pseudocode is an appendix.  I'm ok with the pseudocode not being perfect, especially for cases that are fairly unlikely in practice.

Reply:
The point of the pseudo-code is not to tell implementations exactly what to do. You can still implement it any way you want.
I'd like to avoid inconsistencies between the text and the pseudo code. We had these before when we introduced packet number spaces, and it now turns out to be quite a lot of work to get things straight.